set.mm

```diff
@@ -373209,6 +373360,149 @@ minus one divided by eight (` ( 2 /L N ) ` = -1^(((N^2)-1)/8) ).
+      2lt4 modid mp4an 1ne2 eqcom com12 syl5bi syl6bi impcomd expd com34 eqcoms
+      eqneqall com14 imp31 sylbid expimpd pm2.61ine pm4.71d nn0re syl2an bibi2d
+      2a1 ltlen mpbird ex pm5.32rd reubidva mpbid ) ADEZAFGHZIJZKZBUAZCUAZUBLZX
+      NMNHZXOMNHZOHZAJZKZCPQZBPQXNXOUCLZXTKZCPQZBPQABCUDXMYBYEBPXMXNPEZKZYAYDCP
+      YGXOPEZKZXTXPYCYIXTXPYCRZYIXTKZYJXPXPXOXNUNZKZRZYKXPYLYKXPYLSZSXOXNCBUOZY
+      IXTYOYPYIKZXTMXQTHZAJZYOYQXSYRAYQXSXQXQOHZYRYPXSYTJYIYPXRXQXQOXOXNMNUEUFU
+      GYGYTYRJZYPYHYFUUAXMYFXQUHEZUUAYFXNXNUIUJUUBYRYTXQUKULUPUMUQURUSYGYSYOSZY
+      PYHXJXLYFUUCYSXLYFXJYOXLYFXJYOSSSAYRAYRJZXLXJYFYOUUDXLXJYFYOSZUUDXLXJKYRF
+      GHZIJZYRDEZKZUUEUUDXLUUGXJUUHUUDXKUUFIAYRFGUEUSAYRDUTVAYFUUIYOYFUUHUUGYOY
+      FUUHXQIJZUUGYOSYFXQVFEZUUHUUJRYFXNVFEMPEUUKXNVBVCXNMVGVDXQVEUPUUJUUGMIJZY
+      OUUJUUFMIUUJUUFMFGHZMUUJYRMFGUUJYRMITHMXQIMTVHVIVJVKMVLEFVMEZVNMUBLMFUCLU
+      UMMJVOFVPEUUNVQFVRVSVTWAMFWBWCVJUSIMUNZUULYOSWDUULIMJZUUOYOMIWEUUPUUOYOYO
+      IMWMWFWGVSWHWHWIWFWHWJWKWLWNWOUQWPWQYLYKXPXCWRWSYIYJYNRXTYIYCYMXPYGXNVLEZ
+      XOVLEYCYMRYHYFUUQXMXNWTUMXOWTXNXOXDXAXBUGXEXFXGXHXHXI $.
+
+    $( There exists a unique decomposition of a prime as a sum of squares of
+       two different nonnegative integers iff ` P == 1 ` (mod 4).  (Contributed
+       by AV, 10-Jun-2023.)  The prime needs not be odd, as observed by WL.
+       (Revised by AV, 18-Jun-2023.) $)
+    2sqreultb $p |- ( P e. Prime
+                      -> ( ( P mod 4 ) = 1 <-> E! a e. NN0 E! b e. NN0
+                              ( a < b /\ ( ( a ^ 2 ) + ( b ^ 2 ) ) = P ) ) ) $=
+      ( cprime wcel co c1 wceq cv clt wbr c2 wa cn0 ex wi wrex adantl cz reximi
+      c4 cmo cexp caddc wreu 2sqreult csn 2reu2rex elsni wb eqeq2 anbi2d breq12
+      2sq2 1re ltnri pm2.21i syl6bi impcomd adantr sylbid rexlimivv syl2imc a1d
+      com23 wn cdif eldif eldifsnneq wo wss nn0ssz id eqcomd ssrexv ax-mp mpsyl
+      3syl eldifi 2sqb syl mpbid ord mpid sylbir expcom pm2.61i impbid ) ADEZAU
+      AUBFGHZBIZCIZJKZWKLUCFWLLUCFUDFZAHZMZCNUEBNUEZWIWJWQABCUFOALUGZEZWIWQWJPZ
+      PWSWTWIWQWPCNQZBNQZWSALHZWJWPBCNNUHZALUIWPXCWJPBCNNWKNEWLNEMZXCWPWJXEXCWP
+      WJPXEXCMWPWMWNLHZMZWJXCWPXGUJXEXCWOXFWMALWNUKULRXEXGWJPXCXEXFWMWJXEXFWKGH
+      WLGHMZWMWJPWKWLUNXHWMGGJKZWJWKGWLGJUMXIWJGUOUPUQURURUSUTVAOVEVBVCVDWIWSVF
+      ZWTWIXJMADWRVGEZWTADWRVHXKWQXCVFZWJADLVIXKWQXLWJPXKWQMZXCWJXMAWNHZCSQZBSQ
+      ZXCWJVJZWQXPXKNSVKZWQXOBNQZXPVLWQXBXNCNQZBNQXSXDXAXTBNWPXNCNWOXNWMWOWNAWO
+      VMVNRTTXTXOBNXRXTXOPVLXNCNSVOVPTVRXOBNSVOVQRXMWIXPXQUJXKWIWQADWRVSUTBCAVT
+      WAWBWCOWDWEWFWGWH $.
+
+    $( There exists a unique decomposition of a prime of the form ` 4 k + 1 `
+       as a sum of squares of two positive integers.  See ~ 2sqnn for the
+       existence of such a decomposition.  (Contributed by AV, 11-Jun-2023.) $)
+    2sqreunn $p |- ( ( P e. Prime /\ ( P mod 4 ) = 1 )
+                     -> E! a e. NN E! b e. NN ( a <_ b
```

Comment:
The double E! operator is difficult, since there might be another a  such that even two or more b exist with the given property, see the comment to df-eu.  I think you meant something closer to the antecedent in 2exeu.  This holds likewise for 2sqreult, 2sqreunnlt, 2sqreunnltb.

Reply:
@wlammen You are absolutely right: The comment does not reflect the theorem, and the pattern "`E! x e. A E! y e. B ph`" cannot be used to express "there exists exactly one ` x ` and one ` y ` such that ` ph ` " (as mentioned in the comment of df-eu). Furthermore, there is no real usage of such a pattern (as well as "E! x E! y ph") in set.mm: only theorems about existential uniqueness itself contain it. Therefore, a different approach must be followed here (as you proposed already).

To use ordered pairs (` E! p e. ( NN X. NN )` ) is possible in this case, because we can assume that it is unique ( by claiming ``( 1st p )` <_ (2nd ` p )``). Alternatively, (proper) unordered pairs can be used: `` E! p e ~P NN ( ( # ` p ) = 2 /\ ... )``, or, using the definition of proper pairs in my mathbox:  ``E! p e. ( PrPairs ` NN ) ... ``. I used this already showing that a line passing through a point within a circle around the origin intersects the circle at exactly two different points (see ~inlinecirc02preu.

I will revise the corresponding theorems accordingly (although, the pattern "` E! x e. A E! y e. B ph `" looks nicer), prefering the ordered pair variant.

Reply:
Good catch!

If we need this a lot, maybe a new syntax like `E!! x e. A , y e. B ph` could be of use, which could be defined as in `~2exeu`.
Otherwise, from a pair point of view, something like `E! p e. ( NN X. NN ) ( p = <. a , b >. /\ ph )` would be more readable, especially if `a` and `b` repeat a lot in `ph`.

Reply:
A more general approach to uniqueness takes tuples from a multi-dimensional space (with possibly differently typed dimensions, like one being the real line, the  second a finite set, and so on), and claims there is just one with a given property.  Perhaps you can take this into account when deciding on the type of a pair.